### PR TITLE
spanconfig: consult job lease interval when updating span configs 

### LINKED ISSUE #73789: spanconfig: ensure safety without mutual exclusion in the reconciliation job
**Describe the problem**

We introduced a singleton `AUTO SPAN CONFIG RECONCILATION` job in #68522. We use this per-tenant job to drive the reconciliation process between a tenant's SQL config state (zone configs) with the cluster's KV state (span configs) -- more details found in the [accompanying RFC](#66348). The RFC relies on strict mutual exclusion guarantees, needing to ensure that at a given point in time, only a single reconciliation instance is up and running. We're currently failing to provide this guarantee as described in the thread here: https://github.com/cockroachdb/cockroach/pull/71994#pullrequestreview-831032004. This issue tracks fixing this bug.

Jira issue: CRDB-11751

### PR DESCRIPTION
**Pull request by arulajmani**
Span config reconciliation is driven by the auto span config
reconciliation job which builds over our existing jobs infrastructure.
Jobs can be claimed by any node in the system. The jobs subsystem uses
sqlliveness to claim jobs and a node has a disjoint lease over a claimed
job until its session expires. This means transactions on behalf of this
job must commit before the session expires if they wish to make strong
mutual exclusion guarantess. The auto span config reconciliation job
relies on such mutual exclusion guarantees.

This patch changes the UpdateSpanConfigRequest to now include lease
information in its proto. Typically, this is the lease interval of the
span config job issuing these requests. Downstream, the KVAccessor
ensures the transaction used to perform the update commits at a
timestamp within this interval.

Release note: None

### CODE REVIEW

**Review by ajwerner — CHANGES_REQUESTED** [review]
Here's some comments. Didn't make it all the way through. 

> Reviewed 4 of 4 files at r1, 14 of 32 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-:-N-1KhDC2jDMbfiaFIO9:b-ddxjgc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, @nvanbenschoten, and @shermanCRL)

___
*[pkg/migration/migrations/seed_tenant_span_configs.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1ImYrEft8YdmZWNmO:-N-1ImYrEft8YdmZWNmP:bsou8y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/migration/migrations/seed_tenant_span_configs.go#L89)):*
> ```Go
> 			}
> 			if err := scKVAccessor.UpdateSpanConfigRecords(
> 				ctx, nil, toUpsert, hlc.MinTimestamp, hlc.MaxTimestamp,
> ```

nit: add back comment or make a var that you set to nil 
___
*[pkg/roachpb/span_config.proto, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1Ith664vytPEFHVJs:-N-1Ith664vytPEFHVJt:bcn85wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/roachpb/span_config.proto#L320)):*
> ```protobuf
> // the delete+upsert semantics described above.
> //
> // The delete/upsert updates are performed transactionally and the timestamp of
> ```

Is it important or relevant that the updates are performed transactionally? Would it break anything if, internally, more than one transaction were used, but all of the transactions carried the supplied deadline?
___
*[pkg/spanconfig/spanconfig.go, line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1JCpQAi-KYzx60xny:-N-1JCpQAi-KYzx60xnz:b-gagdry) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfig.go#L195)):*
> ```Go
> 	// [1]: It's possible for system.{zones,descriptor} to have been GC-ed away;
> 	//      think suspended tenants.
> 	// TODO(arul): write words about the session being passed in.
> ```

nit: do this in this PR. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1JVTt3iXLGp19JlTr:-N-1JVTt3iXLGp19JlTs:b-zdx3bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L38)):*
> ```Go
> // NB: don't change the string here; this will cause cross-version issues since
> // this singleton is used as a marker.
> var errMarkCanRetryInvalidLease = errors.New("should retry with valid lease start time")
> ```

would it be better to use a type and then register that type with the errors framework? That provides a path for migration. I don't know how much better it is. Does this definitely work across RPC boundaries, even in the face of wrapping? I'd love to see a test where the accessor implementation is accessed via a connector. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1KHJX-PKEhPOlzf0D:-N-1KHJX-PKEhPOlzf0E:b-7o2692) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L290)):*
> ```Go
> 		txnCommitTimestamp = k.knobs.KVAccessorOverrideUpdateTransactionTimestamp
> 	}
> 	if leaseExpirationTime.Less(txnCommitTimestamp) {
> ```

this is not what you're looking for. You want: https://github.com/cockroachdb/cockroach/blob/1d4a1a3563d148e0008fb76ba6047210c573172e/pkg/kv/txn.go#L782-L788
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1KV5n-MZD4wbAeLez:-N-1KV5n-MZD4wbAeLf-:b-5zlmeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L309)):*
> ```Go
> 		)
> 	}
> 	if txnCommitTimestamp.Less(leaseStartTime) {
> ```

this also isn't quite right. You want to just do the checking before you even create the transaction. I'd expect that you'd read the hlc, and if it's in the past of the timestamp, wait until it is not.

**Review by arulajmani — COMMENTED** [review]
RFAL 

> Dismissed @ajwerner from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-:-N-VUJqbEH7tGyPmUMmi:b-828fno)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @nvanbenschoten, and @shermanCRL)

___
*[pkg/roachpb/span_config.proto, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1Ith664vytPEFHVJs:-N-VLPab6jbcQpFeLBVN:b-fpgyy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/roachpb/span_config.proto#L320)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is it important or relevant that the updates are performed transactionally? Would it break anything if, internally, more than one transaction were used, but all of the transactions carried the supplied deadline?
</blockquote></details>

No, I don't think so -- "atomically" is probably a better word here, switched.
___
*[pkg/spanconfig/spanconfig.go, line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1JCpQAi-KYzx60xny:-N-VSu13BHardHQJ0XVw:b-49xt2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfig.go#L195)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: do this in this PR. 
</blockquote></details>

oops, done. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1JVTt3iXLGp19JlTr:-N-VSvmH2wn-jJsh2UEK:b-o9kywj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L38)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

would it be better to use a type and then register that type with the errors framework? That provides a path for migration. I don't know how much better it is. Does this definitely work across RPC boundaries, even in the face of wrapping? I'd love to see a test where the accessor implementation is accessed via a connector. 
</blockquote></details>

Good call, this didn't work across RPC boundaries. Fixed. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1KHJX-PKEhPOlzf0D:-N-UCmnQ6LOTvtBMa45y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L290)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is not what you're looking for. You want: https://github.com/cockroachdb/cockroach/blob/1d4a1a3563d148e0008fb76ba6047210c573172e/pkg/kv/txn.go#L782-L788
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1KV5n-MZD4wbAeLez:-N-UCbqWByBt4CFztBjT:b83vpdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L309)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this also isn't quite right. You want to just do the checking before you even create the transaction. I'd expect that you'd read the hlc, and if it's in the past of the timestamp, wait until it is not. 
</blockquote></details>

I wasn't sure if we wanted to wait things out or return errors if we weren't in the given interval initially; switched.

File: pkg/spanconfig/spanconfigkvaccessor/kvaccessor_test.go [hunk 1]
```diff
@@ -276,3 +295,42 @@ span [j,k)
 		require.Equal(t, 2, batches)
 	}
 }
+
+// TestKVAccessorUpdateLeastStartWaitRespondsToContextCancellation ensures that
+// KVAccessor update requests which are waiting for their lease start times to
+// be "valid" respond to context cancellations.
+func TestKVAccessorUpdateLeaseStartWaitRespondsTo(t *testing.T) {
```
Comment (ajwerner) [inline pkg/spanconfig/spanconfigkvaccessor/kvaccessor_test.go:302]: name got trunctated

**Review by ajwerner — APPROVED** [review]
LGTM

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 3 of 4 files at r1, 33 of 33 files at r3, 39 of 39 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-:-N-x3GQ792yMVMH-vN6T:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @miretskiy)

___
*[pkg/roachpb/span_config.proto, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1Ith664vytPEFHVJs:-N-x0etP3ymZOEHJtBmZ:b-mf6dle) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/roachpb/span_config.proto#L320)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

No, I don't think so -- "atomically" is probably a better word here, switched.
</blockquote></details>

Is it possible to guarantee atomicity with multiple txns? If half of the update commits, then time advances (or the spanconfig rows are read) such that the other half cannot commit within the lease interval, what can we do?
___
*[pkg/roachpb/span_config.proto, line 324 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-x0R6K8dT0YKRMdVQ2:-N-x0R6K8dT0YKRMdVQ3:b7y29zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/roachpb/span_config.proto#L324)):*
> ```protobuf
> // [leaseStartTime, leaseExpirationTime) supplied. Typically, this corresponds
> // to the lease interval of the auto span config reconciliation job issuing the
> // request.
> ```

nit: mention what happens if committing the txn within this interval is not possible.
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-x1ZPb3DhgdxDacUcW:-N-x1ZPb3DhgdxDacUcX:b-w9m3jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L180)):*
> ```Go
> 	timer := timeutil.NewTimer()
> 	for {
> 		timer.Reset(leaseStartTime.GoTime().Sub(k.clock.Now().GoTime()))
> ```

Can this use `hlc.Clock.SleepUntil`? That function does its best to avoid busy-looping and is also context aware.
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-x2g3gAFH8W4x7IGSX:-N-x2g3gAFH8W4x7IGSY:b-isov29) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigreconciler/reconciler.go#L406)):*
> ```Go
> 	for retrier := retry.StartWithCtx(ctx, retryOpts); retrier.Next(); {
> 		err := kvAccessor.UpdateSpanConfigRecords(
> 			ctx, toDelete, toUpsert, session.StartTimestamp(), session.Expiration(),
> ```

Do we expect the `Expiration` to implicitly advance from call to call? That's needed for this retry loop to be effective. If so, could we add a comment somewhere?

**Review by arulajmani — COMMENTED** [review]
TFTRs! 

bors r+ 

> Dismissed @ajwerner and @irfansharif from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-:-N-xg0-yD38KzPnayUxR:bp7z2nm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @arulajmani, @irfansharif, @miretskiy, and @nvanbenschoten)

___
*[pkg/jobs/jobs.go, line 52 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-MzWyUjn7-YMLxehvUp4-r4-52:-N-xHfcm8BYOFhB5V6CD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/jobs/jobs.go#L52)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Accidental typo?
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-1Ith664vytPEFHVJs:-N-xQZkW5eF896fHZuGC:bablbc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5678cc8b289b4aa63ca646cc92d93b2da5263a1d/pkg/roachpb/span_config.proto#L320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it possible to guarantee atomicity with multiple txns? If half of the update commits, then time advances (or the spanconfig rows are read) such that the other half cannot commit within the lease interval, what can we do?
</blockquote></details>

You're right, I don't think we'd be able to use more than one transactions here in the future given our atomicity guarantee.  
___
*[pkg/roachpb/span_config.proto, line 322 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-MzWyUjn7-YMLxehvUpE-r4-322:-N-xHrrs1WrHyHngfmOQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/roachpb/span_config.proto#L322)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Use the capitalized "Lease{Start,Expiration}Time" in the comment, easier to grep for.
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfig.go, line 58 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-MzWyUjn7-YMLxehvUpG-r4-58:-N-xPRZbDoyqxIH48WDA:b-hoitby) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfig.go#L58)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Drop "lease" phrasing here, it could instead be "txn{Start,End} hlc.Timestamp" or "commit{Min,Max}" etc. At the level of the interface, dropping references to specific impls (like the reconciliation job) makes things clearer -- you don't don't need to know at this level where these timestamps are coming from and that they are tied to lease durations. In our tests and other usages (SeedSpanConfigs) for e.g. they're not tied to leases at all.
</blockquote></details>

I don't think it's that bad, given the "typically" phrasing. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-MzWyUjo9F7kVpZlTsAj-r4-180:-N-xJWer0aCRMvYf-16M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L180)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you refactor this logic to be like: 
```go
if err := waitForLeaseStart(ctx, k.clock, leaseStartTime); err != nil {
    return err
}
return k.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
	// Set the deadline of the transaction so that it commits before the
	// lease of the job on behalf of which the KVAccessor is acting
	// expires.
	if leaseExpirationTime.Less(txn.ReadTimestamp()) {
		return newRetryableLeaseExpiredError()
	}
	if err := txn.UpdateDeadline(ctx, leaseExpirationTime); err != nil {
		return err
	}
	return k.updateSpanConfigRecordsWithTxn(ctx, toDelete, toUpsert, txn)
})
```
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-x1ZPb3DhgdxDacUcW:-N-xJ6wY9rEp9ifQn4sY:baf33ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor.go#L180)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can this use `hlc.Clock.SleepUntil`? That function does its best to avoid busy-looping and is also context aware.
</blockquote></details>

Good call, done. 
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 406 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-N-x2g3gAFH8W4x7IGSX:-N-xJmv41XM1CXwAkK6s:b-mw5hso) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigreconciler/reconciler.go#L406)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we expect the `Expiration` to implicitly advance from call to call? That's needed for this retry loop to be effective. If so, could we add a comment somewhere?
</blockquote></details>

Yup, whenever the underlying sqlliveness session is extended. Added some words. 
___
*[pkg/spanconfig/spanconfigkvaccessor/kvaccessor_test.go, line 302 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-MzXDSBeILMMqX-5jsls-r4-302:-N-xKg2_7v1D9OHE-J_r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71070e7883c39661aba200922e114f3d716d992d/pkg/spanconfig/spanconfigkvaccessor/kvaccessor_test.go#L302)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

name got trunctated
</blockquote></details>

Done.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 40 of 40 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79171#-:-N-xyACp8VfiCEfxrRTE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @miretskiy)

**Comment by arulajmani** [comment]
blathers backport release-22.1

**Pull request merged**